WalletWasabi/Wabisabi/WabiSabiClient.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Crypto.ZeroKnowledge.NonInteractive;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Wabisabi
+{
+	public class WabiSabiClient
+	{
+		public WabiSabiClient(CoordinatorParameters coordinatorParameters, int numberOfCredentials, WasabiRandom rnd)
+		{
+			Random = Guard.NotNull(nameof(rnd), rnd);
+			NumberOfCredentials = Guard.InRangeAndNotNull(nameof(numberOfCredentials), numberOfCredentials, 1, 100);
+			CoordinatorParameters = Guard.NotNull(nameof(coordinatorParameters), coordinatorParameters);
+		}
+
+		private CoordinatorParameters CoordinatorParameters { get; }
+
+		private WasabiRandom Random { get; }
+
+		private int NumberOfCredentials { get; }
+
+		public CredentialPool Credentials { get; } = new CredentialPool();
+
+		public CredentialRegistrationFactory GetCredentialRegistrationFactory()
+			=> new CredentialRegistrationFactory(NumberOfCredentials, Credentials, CoordinatorParameters, Random);
+
+		public void HandleResponse(RegistrationResponse registrationResponse, RegistrationValidationData registrationValidationData)
+		{
+			Guard.NotNull(nameof(registrationResponse), registrationResponse);
+			Guard.NotNull(nameof(registrationValidationData), registrationValidationData);
+
+			var issuedCredentialCount = registrationResponse.IssuedCredentials.Count();
+			var requestedCredentialCount = registrationValidationData.Requested.Count();
+			if (issuedCredentialCount != NumberOfCredentials)
```

Comment:
should not this be `if (issuedCredentialCount != requestedCredentialCount)`

Reply:
No. The `NumberOfCredentials` is a protocol level constant (this is called `k` in the WabiSabi paper)